### ISSUE TITLE
Is there a business case for a GIS index?

### ISSUE THREAD

**Issue by dimaqq** [scylladb/scylla#9940]
It seems that today, if someone needs an efficient GIS/geometry (either lat,lon or x,y) lookup in a general-purpose database, the only (?🗡️) choice is the PostGIS extension for PostgreSQL.

🗡️
Someone made an overview: file:///home/dima/Downloads/ijgi-09-00331-v2.pdf it shows some nosql db features actually exist)
Apparently there is a plugin for Cassandra.

There doesn't appear to be an extension like that for any (?) NoSQL databases.

I wonder, if, perhaps there is a use case to implement this functionality in Scylla.

PostGIS provides a whole lot more than indexing, but when it comes to indices, it seems that the key features are:
- 2d coordinates
- 3d coordinates
- index cardinality analysis for query
- bounding box operators
- distance operators

A related standard is GeoJSON, which provides:
- primitives: point / linestring / polygon
- multipart of the above: multi- (point / linestring / polygon) and geometrycollection (bag of any of the above?)
- operators (intersects, within, contains, within_distance, etc.) (library features, strictly speaking not a part of standard)

A DIY solution usually is to implement https://en.wikipedia.org/wiki/Geohash (square grid) or https://eng.uber.com/h3/ (hexagonal grid).
These are more generic and require a lot less code in the server, offloading "operators" to the client.
Perhaps that's something that's likelier to land in Scylla?

Then there's the whole "raster data" vs. "vector data" split, which makes the above not exhaustive at all 😓

**Comment by dorlaor** [comment]
On Tue, Jan 18, 2022 at 4:11 AM Dima Tisnek ***@***.***>
wrote:

> It seems that today, if someone needs an efficient GIS/geometry (either
> lat,lon or x,y) lookup in a general-purpose database, the only (?🗡️)
> choice is the PostGIS extension for PostgreSQL.
>
> 🗡️
> Someone made an overview: file:///home/dima/Downloads/ijgi-09-00331-v2.pdf
> it shows some nosql db features actually exist)
> Apparently there is a plugin for Cassandra.
>
> There doesn't appear to be an extension like that for any (?) NoSQL
> databases.
>
> I wonder, if, perhaps there is a use case to implement this functionality
> in Scylla.
>
> PostGIS provides a whole lot more than indexing, but when it comes to
> indices, it seems that the key features are:
>
>    - 2d coordinates
>    - 3d coordinates
>    - index cardinality analysis for query
>    - bounding box operators
>    - distance operators
>
> A related standard is GeoJSON, which provides:
>
>    - primitives: point / linestring / polygon
>    - multipart of the above: multi- (point / linestring / polygon) and
>    geometrycollection (bag of any of the above?)
>    - operators (intersects, within, contains, within_distance, etc.)
>    (library features, strictly speaking not a part of standard)
>
> A DIY solution usually is to implement
> https://en.wikipedia.org/wiki/Geohash (square grid) or
> https://eng.uber.com/h3/ (hexagonal grid).
> These are more generic and require a lot less code in the server,
> offloading "operators" to the client.
> Perhaps that's something that's likelier to land in Scylla?
>
Spatial info is a good match to Scylla and was requested by other users in
the past.
The question is whether Scylla can be extended in a minimal way to support
it.
Maybe a set of user defined functions (through WASM) and ability to index
by a such
a function will be enough? Later on those user defined functions can form
an official api.
Most of the logic will be on the client side, using one of the existing
algorithms/hashes
As this is quite some work, we'll need developers like you to drive it

**Comment by dimaqq** [comment]
Index over a user-defined function over field/item makes sense.
I think Mongo had that model w/ javascript.

Actually if client has that function, they can store the result as another field already.
That's DynamoDB spacial index approach.

What I think is missing are the different operators to compare index values to query values.
Prefix search is not enough.
I think some solutions utilise specific kind of tree to store the index, which is out of scope of a user function, isn't it?

Re developers like me:
I'm recently in a new role, and don't have time...
Yet, if some db (e.g. redis, apparently) provides minimal support for spacial data, I will try to steer my current employer from "RDBMS because Postgre because it has PostGIS" to "NoSQL all the way, using barebones solution X".

**Comment by dimaqq** [comment]
Re efficient index:
A common query is "give me all objects in a certain radius" or "all objects in my viewport (certain rectangle)".
In case of geohash (squares or hex), the current point may be close to the border of a macrocell, which means a single prefix lookup will not do.
Rather a typical lookup will be: identify cell of a good size (smaller than query) for current point [client-side], find all neighbours of the same size that overlap the disc/rectangle of interest [client-side], issue a query to the DB for several prefixes at once:

* current point, e.g. `4456123789`
* current point's cell, e.g. `4456`
* neighbours: `4455, 4457, 4458, 4155, 4156, 4157, 4158` (from `44*` block and `41*` block)
* simpler query `geohash ~= 445* or geohash ~= 415*` (2 prefixes)
* or more exact query `geohash ~= 4455* or geohash ~= 4155* or ...` (6 prefixes)

SQL databases are typically pretty bad at handling the `or` queries, and that makes sense in the general case (or across different indices).
My guess is that supporting geohash means supporting `or` efficiently when the branches are all prefixes on the same field/index.

**Comment by JonasHiltl** [comment]
@dimaqq I totally agree with your propsed solution to offload the encoding of lat/lang to the client, which would be easier and faster to implement for Scylla. 
For basic geospatial queries, like "all objects in a given radius" Scylla does not need specific geo functions just efficiant `or` queries with prefix search on indexes. But based on this [issue](https://github.com/scylladb/scylla/issues/4477#issuecomment-525358891) prefix search is only possbile through regular filtering which uses a full table scan.

**Comment by PeterCorless** [comment]
There are currently two different strategies to add geospatial queries to Cassandra: 

- [GeoMesa](https://www.geomesa.org/) — a package that would reside on top of ScyllaDB; it would be using ScyllaDB as an underlying datastore, much the way that JanusGraph can reside on top of ScyllaDB for a graph data model. This is available today. GeoMesa uses Z-curve indexes. GeoMesa uses Contextual Query Language (CQL), which should not be confused with Cassandra Query Language (CQL). It's been described as "heavyweight" but if you want to start today, this is the most facile way to get going.

- Native geospatial support. For example, adding R*-trees indexes and a geospatial query planner to ScyllaDB, as per the design of [this 2019 paper](https://scholarworks.sjsu.edu/cgi/viewcontent.cgi?article=1714&context=etd_projects) which added R*-trees to Cassandra. The R*-tree outperforms Z-curve indexes for multidimensional data, and can be used for K Nearest Neighbor (KNN) analysis in a way that Z-curves cannot support. This paper does not get as far as many types of common geospatial operations, like defining polygons, lines, pathing and bounding boxes.
